lib/filterx/object-metrics-labels.c

```diff
@@ -0,0 +1,352 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "object-metrics-labels.h"
+#include "object-dict-interface.h"
+#include "object-string.h"
+#include "object-primitive.h"
+#include "object-extractor.h"
+#include "filterx-object-istype.h"
+#include "filterx-ref.h"
+#include "filterx-eval.h"
+#include "stats/stats-prometheus.h"
+#include "stats/stats-cluster.h"
+#include "scratch-buffers.h"
+
+#include <string.h>
+
+#define DEDUP_METRICS_LABELS_USAGE "Usage: dedup_metrics_labels(my_metrics_labels)"
+
+typedef struct _FilterXObjectMetricsLabels
+{
+  FilterXDict super;
+  GArray *labels;
+  gboolean sorted;
+  gboolean deduped;
+} FilterXObjectMetricsLabels;
+
+static gboolean
+_truthy(FilterXObject *s)
+{
+  return TRUE;
+}
+
+static gboolean
+_marshal(FilterXObject *s, GString *repr, LogMessageValueType *t)
+{
+  FilterXObjectMetricsLabels *self = (FilterXObjectMetricsLabels *) s;
+
+  stats_prometheus_format_labels_append((StatsClusterLabel *) self->labels->data, self->labels->len, repr);
+  if (t)
+    *t = LM_VT_STRING;
+  return TRUE;
+}
+
+static gboolean
+_repr(FilterXObject *s, GString *repr)
+{
+  _marshal(s, repr, NULL);
+  return TRUE;
+}
+
+static guint64
+_len(FilterXDict *s)
+{
+  FilterXObjectMetricsLabels *self = (FilterXObjectMetricsLabels *) s;
+
+  return self->labels->len;
+}
+
+static FilterXObject *
+_get_subscript(FilterXDict *s, FilterXObject *key)
+{
+  FilterXObjectMetricsLabels *self = (FilterXObjectMetricsLabels *) s;
+
+  const gchar *key_str;
+  if (!filterx_object_extract_string_ref(key, &key_str, NULL))
+    return NULL;
+
+  for (guint i = self->labels->len - 1; i >= 0; i--)
+    {
+      StatsClusterLabel *label = &g_array_index(self->labels, StatsClusterLabel, i);
+      if (strcmp(label->name, key_str) == 0)
+        return filterx_string_new(label->value, -1);
+    }
+
+  return NULL;
+}
+
+static const gchar *
+_obj_to_string(FilterXObject *obj, gsize *len)
+{
+  const gchar *str;
+  if (filterx_object_extract_string_ref(obj, &str, len))
+    return str;
+
+  GString *buf = scratch_buffers_alloc();
+  if (filterx_object_repr(obj, buf))
+    {
+      *len = buf->len;
+      return buf->str;
+    }
+
+  return NULL;
+}
+
+static gboolean
+_set_subscript(FilterXDict *s, FilterXObject *key, FilterXObject **new_value)
+{
+  FilterXObjectMetricsLabels *self = (FilterXObjectMetricsLabels *) s;
+
+  ScratchBuffersMarker marker;
+  scratch_buffers_mark(&marker);
+
+  gsize name_len, value_len;
+  const gchar *name_str = _obj_to_string(key, &name_len);
+  const gchar *value_str = _obj_to_string(*new_value, &value_len);
+  if (!name_str || !value_str)
+    return FALSE;
+
+  FilterXObject *typed_new_value = filterx_ref_unwrap_ro(*new_value);
```

Comment:
unwrap is not necessary here (string is immutable).

Reply:
This might not be a string, this is why I introduced the check. Dicts can be passed here as well, after a `repr()` call we will store it, and return the newly allocated repr version of the dict.

Reply:
As long as your the type you compare against is immutable, you don't need the unwrap call.

Reply:
You are right, I overlooked that. Fixed :)